src/guide/advanced/ffi.md

```diff
@@ -20,59 +20,66 @@ The main standard used today is the C application binary interface. It's
 simple, it's guaranteed to be available even in languages which can't
 compile to WASM, and it's stable. In principle, you could do everything
 manually, but Iroha provides you with a crate `iroha_ffi` which contains
-all you need to create compliant functions.
+all you need to generate FFI-compliant functions out of your existing `Rust` API. 
+
+Automatic generation of FFI bindings and conversion of types alleviates the user of the mental dilligence required not to induce UB when interfacing `unsafe` code which essentially every FFI function call is. This is achieved by insisting that every type used in the generated extern function API is a **robust** `repr(C)` type. The only notable exception being pointers where, except for the null check, pointer validity cannot be determined.
+
+## Example
+
+```rs
+#[derive(FfiType)]
+struct DaysSinceEquinox(u32);
+
+#[ffi_export]
+impl DaysSinceEquinox {
+    pub fn update_value(&mut self, a: &u8) {
+        self.0 = *a as u32;
+    }
+}
+```
+
+will generate the following binding with `DaysSinceEquinox` represented as an opaque pointer:
+
+```rs
+pub extern fn DaysSinceEquinox__update_value(handle: *mut DaysSinceEquinox, a: *const u8) -> FfiReturn {
+    // function implementation
+}
+```
 
 ## FFI Binding Generation
 
-The `iroha_ffi` crate is used to generate functions that are callable via
-FFI.
+The `iroha_ffi` crate is used to generate functions that are callable via FFI out of existing `Rust` structs and methods.
 
-A Rust structure is converted into an intermediary FFI struct via the
-`IntoFfi` trait, which implements the `AsReprCRef` trait, and can be
-converted into a type that can cross FFI boundary. This goes the other way
-around as well: FFI ReprC type is converted into a struct via
-`TryFromReprC`.
+A Rust type is converted into a robust `repr(C)` type that can cross FFI boundary with `FfiType::into_ffi`. This goes the other way around as well: FFI `ReprC` type is converted into a `Rust` type via `FfiType::try_from_ffi`. Obviously, the conversion in the other direction is fallible and the library makes the best effort to protect the caller from all possible forms of accidental UB. 
```

Comment:
```suggestion
A Rust type is converted into a robust `repr(C)` type that can cross FFI boundary with `FfiType::into_ffi`. This goes the other way around as well: FFI `ReprC` type is converted into a `Rust` type via `FfiType::try_from_ffi`. 

::: warning 

Note that the opposite conversion is fallible and can cause undefined behaviour. While we can make the best effort to avoid the most obvious mistakes, you must ensure the correctness of the program on your end or suffer the consequences. 
:::
```